docs/proposals/scheduling/20220629-schedule-device-in-card-level.md

```diff
@@ -0,0 +1,141 @@
+like gpu-nvlink in A100\V100、gpu Mig in A100.
+
+### Goals
+1. a device reporter per node will report card level device crd, which will not conflict with kubelet and existed device plugin
+
+2. scheduler schedule pod with device card level
+
+3. runtime-proxy use scheduler device schedule result to overwrite kubelet\device-plugin card assign result
+
+### User Stories
+
+#### Story NVLink
+NVLink
+
+![image](/docs/images/nvlink.jpg)
+
+through the picture we can find that although there's 8 card in A100\V100 node, the data-transport-speed between cards 
+are not same. when a pod need multi-gpu-cards, we have chance to select a better placement with largest data-transport-speed. 
+
+#### Story MIG
+![image](/docs/images/mig.jpg)
+
+A100 gpu card support mig function to split a card into different size of ci\gi, which provide isolate ability.
+in some scenarios, we need create ci\gi dynamically with pod-scheduling, for example, some pods need 1c.4g.40gb, some pods
+need 1c.2g.20gb, we can't expect the pods requirements in advance, so there is a demand for scheduler to make the 
+best placement for ci\gi-create and schedule pod in it.
+
+#### Story Oversold
+in some cases, we find although some cards free is zero in scheduler account, but we observed that the physical
+utilization of gpu-mem\gpu-core still has buffer, so we may schedule some low-priority jobs to use this buffer, and
+when buffer disappear, we should kill these jobs to avoid affecting normal jobs. because the utilization is different
+bewteen cards, so scheduler must schedule based card level.
+     
+
+## Implementation Details
+
+#### Scheduler
+1. we need a new data structure to describe each device card resources and healthy status on one node
+2. we need new plugins to execute filter\reserve\bind stage
+3. support user to apply any ratio of one card, like 50%/25%
+4. automatically recognize different kind devices. when a new device added, we don't need modify any code
+5. schedule-result will be written on pod annotations.
+
+#### device reporter
+we need create a new device reporter on node to update a new device crd to report each device card resource and health status
+
+```go
+type DeviceType string
+
+const (
+	GPU  DeviceType = "gpu-device"
+	FPGA DeviceType = "fpga-device"
+	RDMA DeviceType = "rdma-device"
+)
+
+type DeviceList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items  []Device `json:"items"`
+}
+
+type Device struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	NodeName     string        `json:"nodeName"`
+	DeviceType   DeviceType    `json:"deviceType"`
+	Spec         DeviceSpec    `json:"spec,omitempty"`
+}
+
+type DeviceSpec struct {
+	DeviceInfos     []DeviceInfo    `json:"deviceInfos"`
+    deviceTopology  DeviceTopology  `json:"deviceTopology"`
+}
+
+type DeviceInfo struct {
+	ID        string                       `json:"id"`       // Device UUID
```

Comment:
Can we add NUMA information here? Excluding the GPU, same NUMA Node is a better choice when allocating multiple devices in a container.

Reply:
> 1.Can we add NUMA information here? Excluding the GPU, same NUMA Node is a better choice when allocating multiple devices in a container.

There is already a proposal(#209) that uses NodeResourceTopology to describe NUMA Topology information, including such as memory, CPU and device in NUMA Node. NUMA Topology Alignment policies have  different scheduling behaviors, But It may not be appropriate to consider a certain resource individually.

Reply:
I know that proposal, as you mentioned, NUMA Topology information only has resource quantity information, not resource individual information, there will be certain restrictions.

Reply:
I Get it. Really need NUMA Node related information. I also commented on this question separately.